cmd/client.go

```diff
@@ -0,0 +1,188 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Evan Spendlove (@evanSpendlove)
+//
+// Client.go creates the client for connecting to, and making requests to, the
+// gRPC server of Cloudprober.
+//
+// TODOs:
+// TODO(#38) Add support for secure connection to RPC server with credentials, if supported by Cloudprober.
+// TODO: Include details of these errors (addProbeFromConfig method) in the error message.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/golang/glog"
+	proberpb "github.com/google/cloudprober/prober/proto"
+	cpprobes "github.com/google/cloudprober/probes"
+	configpb "github.com/google/cloudprober/probes/proto"
+	"google.golang.org/grpc"
+)
+
+type CloudproberClient struct {
+	conn      *grpc.ClientConn           // Client connection for opening and closing the connection
+	client    proberpb.CloudproberClient // Cloudprober gRPC client retained between RPC calls
+	clientMux sync.Mutex                 // Mutex lock for making CloudproberClient safe for concurrent use
+}
+
+// NewClient creates, initialises and returns a new CloudproberClient.
+// This is the correct way to create a new instance of CloudproberClient.
+// Parameters:
+// - rpcServer: Pass the rpcServer address and port as a string formatted as: "hostname:port", e.g. "localhost:1234"
+// Returns:
+// - error:
+//	  - nil: No error occurred and the client was successfully initialised.
+//	  - gRPC Error: See https://godoc.org/google.golang.org/grpc/codes for error codes.
+//		-> See https://github.com/grpc/grpc-go/blob/d25c71b54334380ff1febd25d88064b36de44b3c/clientconn.go#L123
+func NewClient(rpcServer string) (*CloudproberClient, error) {
+	conn, err := grpc.Dial(rpcServer, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("GRPC_DIAL_ERROR: %v", err)
+	}
+	return &CloudproberClient{conn: conn, client: proberpb.NewCloudproberClient(conn)}, nil
+}
+
+// initClient establishes an active client connection to the Cloudprober gRPC server.
+// If there is already an active connection, it will do nothing and return nil.
+// If there is not an active connection, it will make one and assign a Cloudprober client
+// to the CloudproberClient.client field.
+// Parameters:
+// - rpcServer: Pass the rpcServer address and port as a string formatted as: "hostname:port", e.g. "localhost:1234"
+// Returns:
+// - error:
+//	  - nil: No error occurred and the client was successfully initialised.
+//	  - gRPC Error: See https://godoc.org/google.golang.org/grpc/codes for error codes.
+//		-> See https://github.com/grpc/grpc-go/blob/d25c71b54334380ff1febd25d88064b36de44b3c/clientconn.go#L123
+func (client *CloudproberClient) initClient(rpcServer string) error {
+	client.clientMux.Lock()
+	defer client.clientMux.Unlock()
+	var err error
+
+	if client.client == nil { // If there is not an active client connection, make one.
```

Comment:
Unless the inline comments are really short, just move them to the line above.

```suggestion
      // If there is not an active client connection, make one.
	if client.client == nil {
```

Reply:
This has been removed due to a refactoring of CloudproberClient towards using a constructor instead of `InitClient()`.